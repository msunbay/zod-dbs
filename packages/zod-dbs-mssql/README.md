# zod-dbs-mssql

Microsoft SQL Server provider for [zod-dbs](https://github.com/msunbay/zod-dbs).

## Installation

```bash
npm install zod-dbs-mssql
```

> Note: This package has a peer dependency on `mssql`. If you don't already have it, install it:

```bash
npm install mssql
```

## Cli Usage

```bash
npm install zod-dbs-cli zod-dbs-mssql

npx zod-dbs --provider mssql
```

## Programmatic Usage

```ts
import { generateZodSchemas } from 'zod-dbs';
import { MsSqlServerProvider } from 'zod-dbs-mssql';

const provider = new MsSqlServerProvider();

await generateZodSchemas({
  provider,
  config: {
    // See zod-dbs documentation for available options
  },
});
```
