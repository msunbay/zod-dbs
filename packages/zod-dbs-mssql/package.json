{
  "name": "zod-dbs-mssql",
  "version": "1.0.0",
  "description": "Microsoft SQL Server provider for zod-dbs",
  "main": "./dist/index.js",
  "types": "./dist/index.d.ts",
  "type": "module",
  "scripts": {
    "build": "rm -rf ./dist && tsc --project tsconfig.build.json",
    "test:unit": "vitest --project unit run",
    "test:types": "tsc --noEmit"
  },
  "author": "msunbay",
  "license": "ISC",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/msunbay/zod-dbs.git",
    "directory": "packages/zod-dbs-mssql"
  },
  "publishConfig": {
    "access": "public"
  },
  "files": [
    "dist"
  ],
  "dependencies": {
    "zod-dbs-core": "workspace:*"
  },
  "devDependencies": {
    "@types/mssql": "^9.1.7",
    "mssql": "^11.0.1"
  },
  "peerDependencies": {
    "mssql": ">=11"
  }
}
