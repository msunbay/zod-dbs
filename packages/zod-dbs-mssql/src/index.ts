export { createClient } from './client.js';
export { MsSqlServerProvider } from './MsSqlServerProvider.js';
export { createProvider } from './provider.js';
export type * from './types.js';
