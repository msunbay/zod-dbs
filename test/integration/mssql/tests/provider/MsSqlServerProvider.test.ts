import { MsSqlServerProvider } from 'zod-dbs-mssql';

import { getProviderConfig } from '../../../utils/context.js';

it('returns raw schema column information', async () => {
  const provider = new MsSqlServerProvider();

  const info = await provider.fetchSchemaInfo({
    ...getProviderConfig(),
    schemaName: 'dbo',
  });

  expect(info).toBeDefined();
  expect(info).toMatchSnapshot('rawColumns');
});

it('returns schema models', async () => {
  const provider = new MsSqlServerProvider();

  const info = await provider.getSchemaInformation({
    ...getProviderConfig(),
    include: ['users'],
  });

  expect(info).toBeDefined();
  expect(info.tables).toBeDefined();

  const userTable = info.tables.find((t) => t.name === 'users')!;
  expect(userTable).toBeDefined();
  expect(userTable.columns).toBeDefined();
  expect(userTable.columns).toHaveLength(7);

  expect(userTable).toMatchSnapshot('userTable');
});
