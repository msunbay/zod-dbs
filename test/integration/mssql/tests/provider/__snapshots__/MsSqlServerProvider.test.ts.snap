// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`returns raw schema column information > rawColumns 1`] = `
[
  {
    "dataType": "nvarchar",
    "defaultValue": undefined,
    "description": undefined,
    "isArray": false,
    "isEnum": false,
    "isNullable": false,
    "isSerial": false,
    "maxLen": 128,
    "name": "optname",
    "schemaName": "dbo",
    "tableName": "MSreplication_options",
    "tableType": "table",
  },
  {
    "dataType": "bit",
    "defaultValue": undefined,
    "description": undefined,
    "isArray": false,
    "isEnum": false,
    "isNullable": false,
    "isSerial": false,
    "maxLen": undefined,
    "name": "value",
    "schemaName": "dbo",
    "tableName": "MSreplication_options",
    "tableType": "table",
  },
  {
    "dataType": "int",
    "defaultValue": undefined,
    "description": undefined,
    "isArray": false,
    "isEnum": false,
    "isNullable": false,
    "isSerial": false,
    "maxLen": undefined,
    "name": "major_version",
    "schemaName": "dbo",
    "tableName": "MSreplication_options",
    "tableType": "table",
  },
  {
    "dataType": "int",
    "defaultValue": undefined,
    "description": undefined,
    "isArray": false,
    "isEnum": false,
    "isNullable": false,
    "isSerial": false,
    "maxLen": undefined,
    "name": "minor_version",
    "schemaName": "dbo",
    "tableName": "MSreplication_options",
    "tableType": "table",
  },
  {
    "dataType": "int",
    "defaultValue": undefined,
    "description": undefined,
    "isArray": false,
    "isEnum": false,
    "isNullable": false,
    "isSerial": false,
    "maxLen": undefined,
    "name": "revision",
    "schemaName": "dbo",
    "tableName": "MSreplication_options",
    "tableType": "table",
  },
  {
    "dataType": "int",
    "defaultValue": undefined,
    "description": undefined,
    "isArray": false,
    "isEnum": false,
    "isNullable": false,
    "isSerial": false,
    "maxLen": undefined,
    "name": "install_failures",
    "schemaName": "dbo",
    "tableName": "MSreplication_options",
    "tableType": "table",
  },
  {
    "dataType": "varchar",
    "defaultValue": undefined,
    "description": undefined,
    "isArray": false,
    "isEnum": false,
    "isNullable": false,
    "isSerial": false,
    "maxLen": 30,
    "name": "xserver_name",
    "schemaName": "dbo",
    "tableName": "spt_fallback_db",
    "tableType": "table",
  },
  {
    "dataType": "datetime",
    "defaultValue": undefined,
    "description": undefined,
    "isArray": false,
    "isEnum": false,
    "isNullable": false,
    "isSerial": false,
    "maxLen": undefined,
    "name": "xdttm_ins",
    "schemaName": "dbo",
    "tableName": "spt_fallback_db",
    "tableType": "table",
  },
  {
    "dataType": "datetime",
    "defaultValue": undefined,
    "description": undefined,
    "isArray": false,
    "isEnum": false,
    "isNullable": false,
    "isSerial": false,
    "maxLen": undefined,
    "name": "xdttm_last_ins_upd",
    "schemaName": "dbo",
    "tableName": "spt_fallback_db",
    "tableType": "table",
  },
  {
    "dataType": "smallint",
    "defaultValue": undefined,
    "description": undefined,
    "isArray": false,
    "isEnum": false,
    "isNullable": true,
    "isSerial": false,
    "maxLen": undefined,
    "name": "xfallback_dbid",
    "schemaName": "dbo",
    "tableName": "spt_fallback_db",
    "tableType": "table",
  },
  {
    "dataType": "varchar",
    "defaultValue": undefined,
    "description": undefined,
    "isArray": false,
    "isEnum": false,
    "isNullable": false,
    "isSerial": false,
    "maxLen": 30,
    "name": "name",
    "schemaName": "dbo",
    "tableName": "spt_fallback_db",
    "tableType": "table",
  },
  {
    "dataType": "smallint",
    "defaultValue": undefined,
    "description": undefined,
    "isArray": false,
    "isEnum": false,
    "isNullable": false,
    "isSerial": false,
    "maxLen": undefined,
    "name": "dbid",
    "schemaName": "dbo",
    "tableName": "spt_fallback_db",
    "tableType": "table",
  },
  {
    "dataType": "smallint",
    "defaultValue": undefined,
    "description": undefined,
    "isArray": false,
    "isEnum": false,
    "isNullable": false,
    "isSerial": false,
    "maxLen": undefined,
    "name": "status",
    "schemaName": "dbo",
    "tableName": "spt_fallback_db",
    "tableType": "table",
  },
  {
    "dataType": "smallint",
    "defaultValue": undefined,
    "description": undefined,
    "isArray": false,
    "isEnum": false,
    "isNullable": false,
    "isSerial": false,
    "maxLen": undefined,
    "name": "version",
    "schemaName": "dbo",
    "tableName": "spt_fallback_db",
    "tableType": "table",
  },
  {
    "dataType": "varchar",
    "defaultValue": undefined,
    "description": undefined,
    "isArray": false,
    "isEnum": false,
    "isNullable": false,
    "isSerial": false,
    "maxLen": 30,
    "name": "xserver_name",
    "schemaName": "dbo",
    "tableName": "spt_fallback_dev",
    "tableType": "table",
  },
  {
    "dataType": "datetime",
    "defaultValue": undefined,
    "description": undefined,
    "isArray": false,
    "isEnum": false,
    "isNullable": false,
    "isSerial": false,
    "maxLen": undefined,
    "name": "xdttm_ins",
    "schemaName": "dbo",
    "tableName": "spt_fallback_dev",
    "tableType": "table",
  },
  {
    "dataType": "datetime",
    "defaultValue": undefined,
    "description": undefined,
    "isArray": false,
    "isEnum": false,
    "isNullable": false,
    "isSerial": false,
    "maxLen": undefined,
    "name": "xdttm_last_ins_upd",
    "schemaName": "dbo",
    "tableName": "spt_fallback_dev",
    "tableType": "table",
  },
  {
    "dataType": "int",
    "defaultValue": undefined,
    "description": undefined,
    "isArray": false,
    "isEnum": false,
    "isNullable": true,
    "isSerial": false,
    "maxLen": undefined,
    "name": "xfallback_low",
    "schemaName": "dbo",
    "tableName": "spt_fallback_dev",
    "tableType": "table",
  },
  {
    "dataType": "char",
    "defaultValue": undefined,
    "description": undefined,
    "isArray": false,
    "isEnum": false,
    "isNullable": true,
    "isSerial": false,
    "maxLen": 2,
    "name": "xfallback_drive",
    "schemaName": "dbo",
    "tableName": "spt_fallback_dev",
    "tableType": "table",
  },
  {
    "dataType": "int",
    "defaultValue": undefined,
    "description": undefined,
    "isArray": false,
    "isEnum": false,
    "isNullable": false,
    "isSerial": false,
    "maxLen": undefined,
    "name": "low",
    "schemaName": "dbo",
    "tableName": "spt_fallback_dev",
    "tableType": "table",
  },
  {
    "dataType": "int",
    "defaultValue": undefined,
    "description": undefined,
    "isArray": false,
    "isEnum": false,
    "isNullable": false,
    "isSerial": false,
    "maxLen": undefined,
    "name": "high",
    "schemaName": "dbo",
    "tableName": "spt_fallback_dev",
    "tableType": "table",
  },
  {
    "dataType": "smallint",
    "defaultValue": undefined,
    "description": undefined,
    "isArray": false,
    "isEnum": false,
    "isNullable": false,
    "isSerial": false,
    "maxLen": undefined,
    "name": "status",
    "schemaName": "dbo",
    "tableName": "spt_fallback_dev",
    "tableType": "table",
  },
  {
    "dataType": "varchar",
    "defaultValue": undefined,
    "description": undefined,
    "isArray": false,
    "isEnum": false,
    "isNullable": false,
    "isSerial": false,
    "maxLen": 30,
    "name": "name",
    "schemaName": "dbo",
    "tableName": "spt_fallback_dev",
    "tableType": "table",
  },
  {
    "dataType": "varchar",
    "defaultValue": undefined,
    "description": undefined,
    "isArray": false,
    "isEnum": false,
    "isNullable": false,
    "isSerial": false,
    "maxLen": 127,
    "name": "phyname",
    "schemaName": "dbo",
    "tableName": "spt_fallback_dev",
    "tableType": "table",
  },
  {
    "dataType": "varchar",
    "defaultValue": undefined,
    "description": undefined,
    "isArray": false,
    "isEnum": false,
    "isNullable": false,
    "isSerial": false,
    "maxLen": 30,
    "name": "xserver_name",
    "schemaName": "dbo",
    "tableName": "spt_fallback_usg",
    "tableType": "table",
  },
  {
    "dataType": "datetime",
    "defaultValue": undefined,
    "description": undefined,
    "isArray": false,
    "isEnum": false,
    "isNullable": false,
    "isSerial": false,
    "maxLen": undefined,
    "name": "xdttm_ins",
    "schemaName": "dbo",
    "tableName": "spt_fallback_usg",
    "tableType": "table",
  },
  {
    "dataType": "datetime",
    "defaultValue": undefined,
    "description": undefined,
    "isArray": false,
    "isEnum": false,
    "isNullable": false,
    "isSerial": false,
    "maxLen": undefined,
    "name": "xdttm_last_ins_upd",
    "schemaName": "dbo",
    "tableName": "spt_fallback_usg",
    "tableType": "table",
  },
  {
    "dataType": "int",
    "defaultValue": undefined,
    "description": undefined,
    "isArray": false,
    "isEnum": false,
    "isNullable": true,
    "isSerial": false,
    "maxLen": undefined,
    "name": "xfallback_vstart",
    "schemaName": "dbo",
    "tableName": "spt_fallback_usg",
    "tableType": "table",
  },
  {
    "dataType": "smallint",
    "defaultValue": undefined,
    "description": undefined,
    "isArray": false,
    "isEnum": false,
    "isNullable": false,
    "isSerial": false,
    "maxLen": undefined,
    "name": "dbid",
    "schemaName": "dbo",
    "tableName": "spt_fallback_usg",
    "tableType": "table",
  },
  {
    "dataType": "int",
    "defaultValue": undefined,
    "description": undefined,
    "isArray": false,
    "isEnum": false,
    "isNullable": false,
    "isSerial": false,
    "maxLen": undefined,
    "name": "segmap",
    "schemaName": "dbo",
    "tableName": "spt_fallback_usg",
    "tableType": "table",
  },
  {
    "dataType": "int",
    "defaultValue": undefined,
    "description": undefined,
    "isArray": false,
    "isEnum": false,
    "isNullable": false,
    "isSerial": false,
    "maxLen": undefined,
    "name": "lstart",
    "schemaName": "dbo",
    "tableName": "spt_fallback_usg",
    "tableType": "table",
  },
  {
    "dataType": "int",
    "defaultValue": undefined,
    "description": undefined,
    "isArray": false,
    "isEnum": false,
    "isNullable": false,
    "isSerial": false,
    "maxLen": undefined,
    "name": "sizepg",
    "schemaName": "dbo",
    "tableName": "spt_fallback_usg",
    "tableType": "table",
  },
  {
    "dataType": "int",
    "defaultValue": undefined,
    "description": undefined,
    "isArray": false,
    "isEnum": false,
    "isNullable": false,
    "isSerial": false,
    "maxLen": undefined,
    "name": "vstart",
    "schemaName": "dbo",
    "tableName": "spt_fallback_usg",
    "tableType": "table",
  },
  {
    "dataType": "datetime",
    "defaultValue": undefined,
    "description": undefined,
    "isArray": false,
    "isEnum": false,
    "isNullable": false,
    "isSerial": false,
    "maxLen": undefined,
    "name": "lastrun",
    "schemaName": "dbo",
    "tableName": "spt_monitor",
    "tableType": "table",
  },
  {
    "dataType": "int",
    "defaultValue": undefined,
    "description": undefined,
    "isArray": false,
    "isEnum": false,
    "isNullable": false,
    "isSerial": false,
    "maxLen": undefined,
    "name": "cpu_busy",
    "schemaName": "dbo",
    "tableName": "spt_monitor",
    "tableType": "table",
  },
  {
    "dataType": "int",
    "defaultValue": undefined,
    "description": undefined,
    "isArray": false,
    "isEnum": false,
    "isNullable": false,
    "isSerial": false,
    "maxLen": undefined,
    "name": "io_busy",
    "schemaName": "dbo",
    "tableName": "spt_monitor",
    "tableType": "table",
  },
  {
    "dataType": "int",
    "defaultValue": undefined,
    "description": undefined,
    "isArray": false,
    "isEnum": false,
    "isNullable": false,
    "isSerial": false,
    "maxLen": undefined,
    "name": "idle",
    "schemaName": "dbo",
    "tableName": "spt_monitor",
    "tableType": "table",
  },
  {
    "dataType": "int",
    "defaultValue": undefined,
    "description": undefined,
    "isArray": false,
    "isEnum": false,
    "isNullable": false,
    "isSerial": false,
    "maxLen": undefined,
    "name": "pack_received",
    "schemaName": "dbo",
    "tableName": "spt_monitor",
    "tableType": "table",
  },
  {
    "dataType": "int",
    "defaultValue": undefined,
    "description": undefined,
    "isArray": false,
    "isEnum": false,
    "isNullable": false,
    "isSerial": false,
    "maxLen": undefined,
    "name": "pack_sent",
    "schemaName": "dbo",
    "tableName": "spt_monitor",
    "tableType": "table",
  },
  {
    "dataType": "int",
    "defaultValue": undefined,
    "description": undefined,
    "isArray": false,
    "isEnum": false,
    "isNullable": false,
    "isSerial": false,
    "maxLen": undefined,
    "name": "connections",
    "schemaName": "dbo",
    "tableName": "spt_monitor",
    "tableType": "table",
  },
  {
    "dataType": "int",
    "defaultValue": undefined,
    "description": undefined,
    "isArray": false,
    "isEnum": false,
    "isNullable": false,
    "isSerial": false,
    "maxLen": undefined,
    "name": "pack_errors",
    "schemaName": "dbo",
    "tableName": "spt_monitor",
    "tableType": "table",
  },
  {
    "dataType": "int",
    "defaultValue": undefined,
    "description": undefined,
    "isArray": false,
    "isEnum": false,
    "isNullable": false,
    "isSerial": false,
    "maxLen": undefined,
    "name": "total_read",
    "schemaName": "dbo",
    "tableName": "spt_monitor",
    "tableType": "table",
  },
  {
    "dataType": "int",
    "defaultValue": undefined,
    "description": undefined,
    "isArray": false,
    "isEnum": false,
    "isNullable": false,
    "isSerial": false,
    "maxLen": undefined,
    "name": "total_write",
    "schemaName": "dbo",
    "tableName": "spt_monitor",
    "tableType": "table",
  },
  {
    "dataType": "int",
    "defaultValue": undefined,
    "description": undefined,
    "isArray": false,
    "isEnum": false,
    "isNullable": false,
    "isSerial": false,
    "maxLen": undefined,
    "name": "total_errors",
    "schemaName": "dbo",
    "tableName": "spt_monitor",
    "tableType": "table",
  },
  {
    "dataType": "nvarchar",
    "defaultValue": undefined,
    "description": undefined,
    "isArray": false,
    "isEnum": false,
    "isNullable": true,
    "isSerial": false,
    "maxLen": 35,
    "name": "name",
    "schemaName": "dbo",
    "tableName": "spt_values",
    "tableType": "table",
  },
  {
    "dataType": "int",
    "defaultValue": undefined,
    "description": undefined,
    "isArray": false,
    "isEnum": false,
    "isNullable": false,
    "isSerial": false,
    "maxLen": undefined,
    "name": "number",
    "schemaName": "dbo",
    "tableName": "spt_values",
    "tableType": "table",
  },
  {
    "dataType": "nchar",
    "defaultValue": undefined,
    "description": undefined,
    "isArray": false,
    "isEnum": false,
    "isNullable": false,
    "isSerial": false,
    "maxLen": 3,
    "name": "type",
    "schemaName": "dbo",
    "tableName": "spt_values",
    "tableType": "table",
  },
  {
    "dataType": "int",
    "defaultValue": undefined,
    "description": undefined,
    "isArray": false,
    "isEnum": false,
    "isNullable": true,
    "isSerial": false,
    "maxLen": undefined,
    "name": "low",
    "schemaName": "dbo",
    "tableName": "spt_values",
    "tableType": "table",
  },
  {
    "dataType": "int",
    "defaultValue": undefined,
    "description": undefined,
    "isArray": false,
    "isEnum": false,
    "isNullable": true,
    "isSerial": false,
    "maxLen": undefined,
    "name": "high",
    "schemaName": "dbo",
    "tableName": "spt_values",
    "tableType": "table",
  },
  {
    "dataType": "int",
    "defaultValue": undefined,
    "description": undefined,
    "isArray": false,
    "isEnum": false,
    "isNullable": true,
    "isSerial": false,
    "maxLen": undefined,
    "name": "status",
    "schemaName": "dbo",
    "tableName": "spt_values",
    "tableType": "table",
  },
  {
    "dataType": "int",
    "defaultValue": undefined,
    "description": undefined,
    "isArray": false,
    "isEnum": false,
    "isNullable": false,
    "isSerial": false,
    "maxLen": undefined,
    "name": "id",
    "schemaName": "dbo",
    "tableName": "users",
    "tableType": "table",
  },
  {
    "dataType": "nvarchar",
    "defaultValue": undefined,
    "description": undefined,
    "isArray": false,
    "isEnum": false,
    "isNullable": false,
    "isSerial": false,
    "maxLen": 100,
    "name": "name",
    "schemaName": "dbo",
    "tableName": "users",
    "tableType": "table",
  },
  {
    "dataType": "nvarchar",
    "defaultValue": undefined,
    "description": undefined,
    "isArray": false,
    "isEnum": false,
    "isNullable": true,
    "isSerial": false,
    "maxLen": 255,
    "name": "email",
    "schemaName": "dbo",
    "tableName": "users",
    "tableType": "table",
  },
  {
    "dataType": "nvarchar",
    "defaultValue": "('active')",
    "description": undefined,
    "isArray": false,
    "isEnum": false,
    "isNullable": true,
    "isSerial": false,
    "maxLen": 20,
    "name": "status",
    "schemaName": "dbo",
    "tableName": "users",
    "tableType": "table",
  },
  {
    "dataType": "nvarchar",
    "defaultValue": undefined,
    "description": undefined,
    "isArray": false,
    "isEnum": false,
    "isNullable": true,
    "isSerial": false,
    "maxLen": -1,
    "name": "profile",
    "schemaName": "dbo",
    "tableName": "users",
    "tableType": "table",
  },
  {
    "dataType": "datetime2",
    "defaultValue": "(sysdatetime())",
    "description": undefined,
    "isArray": false,
    "isEnum": false,
    "isNullable": true,
    "isSerial": false,
    "maxLen": undefined,
    "name": "created_at",
    "schemaName": "dbo",
    "tableName": "users",
    "tableType": "table",
  },
  {
    "dataType": "nvarchar",
    "defaultValue": "('viewer')",
    "description": undefined,
    "isArray": false,
    "isEnum": false,
    "isNullable": true,
    "isSerial": false,
    "maxLen": 20,
    "name": "roles",
    "schemaName": "dbo",
    "tableName": "users",
    "tableType": "table",
  },
]
`;

exports[`returns schema models > userTable 1`] = `
{
  "columns": [
    {
      "dataType": "int",
      "defaultValue": undefined,
      "description": undefined,
      "isArray": false,
      "isEnum": false,
      "isNullable": false,
      "isReadOptional": false,
      "isSerial": false,
      "isWritable": true,
      "isWriteOptional": false,
      "maxLen": undefined,
      "name": "id",
      "schemaName": "dbo",
      "tableName": "users",
      "tableType": "table",
      "type": "int",
    },
    {
      "dataType": "nvarchar",
      "defaultValue": undefined,
      "description": undefined,
      "isArray": false,
      "isEnum": false,
      "isNullable": false,
      "isReadOptional": false,
      "isSerial": false,
      "isWritable": true,
      "isWriteOptional": false,
      "maxLen": 100,
      "name": "name",
      "schemaName": "dbo",
      "tableName": "users",
      "tableType": "table",
      "type": "string",
    },
    {
      "dataType": "nvarchar",
      "defaultValue": undefined,
      "description": undefined,
      "isArray": false,
      "isEnum": false,
      "isNullable": true,
      "isReadOptional": true,
      "isSerial": false,
      "isWritable": true,
      "isWriteOptional": true,
      "maxLen": 255,
      "name": "email",
      "schemaName": "dbo",
      "tableName": "users",
      "tableType": "table",
      "type": "string",
    },
    {
      "dataType": "nvarchar",
      "defaultValue": "('active')",
      "description": undefined,
      "isArray": false,
      "isEnum": false,
      "isNullable": true,
      "isReadOptional": true,
      "isSerial": false,
      "isWritable": true,
      "isWriteOptional": true,
      "maxLen": 20,
      "name": "status",
      "schemaName": "dbo",
      "tableName": "users",
      "tableType": "table",
      "type": "string",
    },
    {
      "dataType": "nvarchar",
      "defaultValue": undefined,
      "description": undefined,
      "isArray": false,
      "isEnum": false,
      "isNullable": true,
      "isReadOptional": true,
      "isSerial": false,
      "isWritable": true,
      "isWriteOptional": true,
      "maxLen": -1,
      "name": "profile",
      "schemaName": "dbo",
      "tableName": "users",
      "tableType": "table",
      "type": "string",
    },
    {
      "dataType": "datetime2",
      "defaultValue": "(sysdatetime())",
      "description": undefined,
      "isArray": false,
      "isEnum": false,
      "isNullable": true,
      "isReadOptional": true,
      "isSerial": false,
      "isWritable": true,
      "isWriteOptional": true,
      "maxLen": undefined,
      "name": "created_at",
      "schemaName": "dbo",
      "tableName": "users",
      "tableType": "table",
      "type": "any",
    },
    {
      "dataType": "nvarchar",
      "defaultValue": "('viewer')",
      "description": undefined,
      "isArray": false,
      "isEnum": false,
      "isNullable": true,
      "isReadOptional": true,
      "isSerial": false,
      "isWritable": true,
      "isWriteOptional": true,
      "maxLen": 20,
      "name": "roles",
      "schemaName": "dbo",
      "tableName": "users",
      "tableType": "table",
      "type": "string",
    },
  ],
  "name": "users",
  "schemaName": "dbo",
  "type": "table",
}
`;
